File: app/services/prosemirror_service/donation_goal_node.rb
# frozen_string_literal: true

module ProsemirrorService
  class DonationGoalNode < ProsemirrorToHtml::Nodes::Node
    @node_type = "donationGoalNode"
    @tag_name = "div"

    def tag
      [{ tag: self.class.tag_name, attrs: (@node.attrs || {}).merge({ class: "donationGoal flex flex-col" }) }]
    end

    def matching
      @node.type == self.class.node_type
    end

    def text
      goal = ProsemirrorService::Renderer.event.donation_goal.progress_amount_cents
      "<p>#{goal.progress_amount_cents} / #{goal.amount_cents}</p><div class=\"bg-black rounded-full w-full\"></div>"
    end

  end
end

File: app/javascript/controllers/tiptap/nodes/donation_goal_node.js
import { Node, mergeAttributes } from '@tiptap/core'

export const DonationGoalNode = Node.create({
  name: 'donationGoal',
  group: 'block',
  priority: 2000,
  renderHTML({ HTMLAttributes }) {
    return [
      'div',
      mergeAttributes(HTMLAttributes, {
        class:
          'donationGoal relative card shadow-none border flex flex-col py-2 my-2',
      }),
      [
        'p',
        { class: 'text-center italic' },
        'Your progress towards your goal will display here',
      ],
      [
        'div',
        { class: 'bg-gray-200 dark:bg-neutral-700 rounded-full w-full' },
        [
          'div',
          {
            class:
              'h-full bg-primary rounded w-1/2 flex items-center justify-center',
          },
          ['p', { class: 'text-sm text-black p-[1px] my-0' }, '50%'],
        ],
      ],
    ]
  },
  parseHTML() {
    return [
      {
        tag: 'div',
        getAttrs: node => node.classList.contains('donationGoal') && null,
      },
    ]
  },
  addCommands() {
    return {
      addDonationGoal:
        () =>
        ({ commands }) => {
          return commands.insertContent({ type: this.name })
        },
    }
  },
})